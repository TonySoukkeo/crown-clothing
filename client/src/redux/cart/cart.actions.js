import CartActionTypes from "./cart.types";

export const toggleCartHidden = () => ({
  type: CartActionTypes.TOGGLE_CART_HIDDEN,
});

export const updateCartStart = (cartTransaction) => ({
  type: CartActionTypes.UPDATE_CART_START,
  payload: cartTransaction,
});

export const updateCartSuccess = (cartItems) => ({
  type: CartActionTypes.UPDATE_CART_SUCCESS,
  payload: cartItems,
});

export const updateCartFailure = (error) => ({
  type: CartActionTypes.UPDATE_CART_FAILURE,
  payload: error,
});

export const clearItemFromCart = (item) => ({
  type: CartActionTypes.CLEAR_ITEM_FROM_CART,
  payload: item,
});

export const clearCart = () => ({
  type: CartActionTypes.CLEAR_CART,
});
