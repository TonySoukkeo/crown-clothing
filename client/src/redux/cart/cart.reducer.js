import CartActionTypes from "./cart.types";

const INITIAL_STATE = {
  hidden: true,
  cartItems: [],
  error: null,
};

const cartReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CartActionTypes.TOGGLE_CART_HIDDEN:
      return {
        ...state,
        hidden: !state.hidden,
      };
    case CartActionTypes.UPDATE_CART_SUCCESS:
      return {
        ...state,
        cartItems: action.payload,
        error: null,
      };

    case CartActionTypes.UPDATE_CART_FAILURE:
      return {
        ...state,
        error: action.payload,
      };

    case CartActionTypes.CLEAR_ITEM_FROM_CART:
      return {
        ...state,
        cartItems: state.cartItems.filter(
          (cartItem) => cartItem.id !== action.payload.id
        ),
        error: null,
      };

    case CartActionTypes.CLEAR_CART:
      return {
        ...state,
        cartItems: [],
        error: null,
      };

    default:
      return state;
  }
};

export default cartReducer;
