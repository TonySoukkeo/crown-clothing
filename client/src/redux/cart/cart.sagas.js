import { all, takeLatest, put, call } from "redux-saga/effects";

import {
  addItemToCart,
  getCartItems,
  removeItemFromCart,
  clearItemFromCart,
} from "./cart.utils";
import UserActionTypes from "../user/user.types";

import {
  clearCart,
  updateCartSuccess,
  updateCartFailure,
} from "./cart.actions";
import CartActionTypes from "./cart.types";
import { firestore } from "../../firebase/firebase.utils";

function* clearCartOnSignOut() {
  yield put(clearCart());
}

function* updateCart({ payload: { currentUser, item, type } }) {
  try {
    const userId = currentUser.id;

    let updatedCartItems = [];

    if (type !== "clear-cart") {
      // Retrieve user cart items from firestore collection
      const cartItems = yield call(getCartItems, userId);

      switch (type) {
        case "add":
          updatedCartItems = yield call(addItemToCart, cartItems, item);
          break;

        case "remove":
          updatedCartItems = yield call(removeItemFromCart, cartItems, item);
          break;

        case "clear-item":
          updatedCartItems = yield call(clearItemFromCart, cartItems, item);
          break;

        default:
          break;
      }
    }

    // Update users cart on firebase with new updated cart
    yield firestore.collection("Carts").doc(currentUser.id).set({
      cartItems: updatedCartItems,
    });

    // Update cart items on client side
    yield put(updateCartSuccess(updatedCartItems));
  } catch (error) {
    yield put(updateCartFailure(error));
  }
}

function* onSignOutSuccess() {
  yield takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

function* onUpdateCart() {
  yield takeLatest(CartActionTypes.UPDATE_CART_START, updateCart);
}

export function* cartSagas() {
  yield all([call(onSignOutSuccess), call(onUpdateCart)]);
}
