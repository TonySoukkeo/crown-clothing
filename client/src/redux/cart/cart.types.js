const CartActionTypes = {
  TOGGLE_CART_HIDDEN: "TOGGLE_CART_HIDDEN",
  ADD_ITEM_START: "ADD_ITEM_START",
  REMOVE_ITEM_START: "REMOVE_ITEM_START",
  CLEAR_ITEM_FROM_CART: "CLEAR_ITEM_FROM_CART",
  CLEAR_CART: "CLEAR_CART",
  UPDATE_CART_START: "UPDATE_CART_START",
  UPDATE_CART_SUCCESS: "UPDATE_CART_SUCCESS",
  UPDATE_CART_FAILURE: "UPDATE_CART_FAILURE",
};

export default CartActionTypes;
