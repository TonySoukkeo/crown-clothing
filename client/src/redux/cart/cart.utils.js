import { firestore } from "../../firebase/firebase.utils";

export const getCartItems = async (userId) => {
  const cart = await firestore.collection("Carts").doc(userId).get();

  // Check if cart exists
  if (!cart.exists) return [];

  return cart.data().cartItems;
};

export const addItemToCart = (cartItems, cartItemToAdd) => {
  const existingCartItem = cartItems.find(
    (cartItem) => cartItem.id === cartItemToAdd.id
  );

  if (existingCartItem) {
    return cartItems.map((cartItem) =>
      cartItem.id === cartItemToAdd.id
        ? { ...cartItem, quantity: cartItem.quantity + 1 }
        : cartItem
    );
  }

  return [...cartItems, { ...cartItemToAdd, quantity: 1 }];
};

export const removeItemFromCart = (cartItems, cartItemToRemove) => {
  const existingCartItem = cartItems.find(
    (cartItem) => cartItem.id === cartItemToRemove.id
  );

  if (existingCartItem.quantity === 1) {
    return cartItems.filter((cartItem) => cartItem.id !== cartItemToRemove.id);
  }

  return cartItems.map((cartItem) =>
    cartItem.id === cartItemToRemove.id
      ? { ...cartItem, quantity: cartItem.quantity - 1 }
      : cartItem
  );
};

export const clearItemFromCart = (cartItems, cartItemToClear) =>
  cartItems.filter((cartItem) => cartItem.id !== cartItemToClear.id);
