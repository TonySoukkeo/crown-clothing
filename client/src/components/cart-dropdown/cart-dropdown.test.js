import CartDropdown from "./cart-dropdown.component";
import { shallow } from "enzyme";
import React from "react";

import { toggleCartHidden } from "../../redux/cart/cart.actions";
import CartItem from "../cart-item/cart-item.component";

import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("CartDropdown component", () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  let mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      history: mockHistory,
      dispatch: mockDispatch,
      cartItems: mockCartItems,
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it("renders CartDropdown withour crashing", () => {
    expect.assertions(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("should call history.push when CartDropdownButton is clicked", () => {
    wrapper.find("CartDropdownButton").simulate("click");

    expect.assertions(2);

    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockHistory.push).toHaveBeenCalledWith("/checkout");
  });

  it("should dispatch correctly when CartDropdownButton is clicked", () => {
    wrapper.find("CartDropdownButton").simulate("click");

    expect.assertions(2);

    expect(mockDispatch).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it("should render out the correct amount of CartItems", () => {
    const cartItem = wrapper.find(CartItem);

    expect.assertions(1);
    expect(cartItem.length).toEqual(mockCartItems.length);
  });

  it("should render EmptyMessageContainer if cartItems is empty", () => {
    const mockProps = {
      cartItems: [],
      history: jest.fn(),
      dispatch: jest.fn(),
    };

    const newWrapper = shallow(<CartDropdown {...mockProps} />);

    expect.assertions(1);
    expect(newWrapper.exists("EmptyMessageContainer")).toEqual(true);
  });
});
