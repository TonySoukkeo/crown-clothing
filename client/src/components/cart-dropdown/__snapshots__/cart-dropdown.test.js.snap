// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component renders CartDropdown withour crashing 1`] = `ShallowWrapper {}`;
